File: leetcode/55.jump_game.rb
'''
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Determine if you are able to reach the last index.

For example:
A = [2,3,1,1,4], return true.

A = [3,2,1,0,4], return false.
'''
require 'set'
def can_jump(nums, visited = Set[])
  n = nums.length
  return true if nums.all?{|x|x>0}
  
  nums.reverse.each_with_index do |num, idx|
    # add every idx that could reach to last of nums
    idx_from_start = n - 1 - idx
    if num >= idx
      visited << idx_from_start
      return true if idx_from_start == 0
    end

    add = false
    visited.each do |idx_v|
      if idx_v > idx_from_start &&  num >= idx_v - idx_from_start
        add = true
        return true if idx_from_start == 0
        break
      end
    end
    visited << idx_from_start if add
  end
  
  visited.include?(0)
end

arr = [3,2,1,0,4]
# arr = [2,5,0,0]
p can_jump(arr)

File: leetcode/pythonSolutions/55.jump_game.py
'''
55. Jump Game
Medium

Share
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Determine if you are able to reach the last index.

Example 1:

Input: [2,3,1,1,4]
Output: true
Explanation: Jump 1 step from index 0 to 1, then 3 steps to the last index.
Example 2:

Input: [3,2,1,0,4]
Output: false
Explanation: You will always arrive at index 3 no matter what. Its maximum
             jump length is 0, which makes it impossible to reach the last index.
'''
class Solution:
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        hash = {}
        hash[0] = True

        for i in range(1, len(nums)):
          for j in range(i-1,-1, -1):
            if j in hash and nums[j] >= i - j:
              hash[i] = True
              break
        return len(nums)-1 in hash
        
nums = [2,3]
print(Solution().canJump(nums))